rpmlb/builder/koji.py

```diff
@@ -0,0 +1,173 @@
+"""Builder interface for the Koji build service."""
+
+import re
+from pathlib import Path
+from typing import Iterator, Optional, Sequence
+
+from ..utils import run_cmd, run_cmd_with_capture
+from ..work import Work
+from .base import BaseBuilder
+
+
+class KojiBuilder(BaseBuilder):
+    """Builder interface for the Koji build service."""
+
+    DEFAULT_TARGET_FORMAT = 'sclo{epel}-{collection}-rh-el{epel}'
+
+    def __init__(self):
+        """Initialize the slots"""
+
+        self.collection = None  #: The collection to build
+        self.epel = None  #: Version of EPEL to build for
+        self.profile = None  #: The koji configuration profile to use
+        #: The format string for construction of the actual template name
+        self.target_format = type(self).DEFAULT_TARGET_FORMAT
+
+    @property
+    def base_command(self) -> Sequence[str]:
+        """Basis of the command for build service interaction."""
+
+        binary = 'koji'
+
+        if self.profile is None:
+            return [binary]
+        else:
+            return [binary, '--profile', self.profile]
+
+    @property
+    def target(self) -> str:
+        """The koji target to build into."""
+
+        return self.target_format.format(
+            collection=self.collection,
+            epel=self.epel,
+        )
+
+    def run(self, work: Work, **kwargs):
```

Comment:
> However, this would mean changing all existing builders and the way the builder is loaded and constructed. That change is bit out of scope of just KojiBuilder…

I can understand what you want to do.
In the template design pattern, if we will follow the design pattern, it is better that `BaseBuilder` class `run` method is a template for all the builder builder sub classes.

Below is my suggestions for that.

> That change is bit out of scope of just KojiBuilder…

Yes, we have to change `BaseBuilder` for my suggestion. But I think that that's fine.

```
    def run(self, work, **kwargs):
        # init or another proper method name.
        init(work, kwargs)
        is_resume = kwargs.get('resume', False)
...

    def init(self, work, **kwargs):
        """Pre-run setup, not skipped on resume.
        Keyword arguments:
            work: Overview of the work to be done.
            koji_profile: The koji configuration profile to use.
            epel: The version of EPEL to build for.
        """
        pass
```

koji.py
```
    def init(self, work, **kwargs):
        self.collection = work._recipe._collection_id
        self.epel = kwargs['epel']
        self.profile = kwargs['koji_profile']
        self.target_format = ( 
            kwargs.get('koji_target_format')
            or type(self).DEFAULT_TARGET_FORMAT
        )
```